### PR TITLE
GEODE-8670: ReconnectDUnitTest is hiding a NullPointerException

### PR DESCRIPTION
**Pull request by bschuchardt**
1. Fixed all of the IgnorableException problems in ReconnectDUnitTest
2. Modified InternalLocator to avoid throwing a NullPointerException and
   added a test for this change

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java [hunk 1]
```diff
@@ -137,6 +137,18 @@ public void startedLocatorHasLocator() throws IOException {
     assertThat(InternalLocator.hasLocator()).isTrue();
   }
 
+  @Test
```
Comment (echobravopapa) [inline geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java:140]: I'm not saying this test is invalid, but it doesn't validate the fix below in `InternalLocator` i.e. if one comments out that code the test still passes...

Comment (bschuchardt) [reply geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java:140]: true, but I couldn't see how to write a better test.  Maybe you'd like to work with me on that?

Comment (bschuchardt) [reply geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java:140]: I think that, in fact, the test is invalid.  I'll post a better one.

**Review by echobravopapa — APPROVED** [review]

File: geode-core/src/distributedTest/java/org/apache/geode/cache30/ReconnectDUnitTest.java [hunk 2]
```diff
@@ -1115,6 +1118,9 @@ public String description() {
    */
   @Test
   public void testReconnectFailsDueToBadCacheXML() throws Exception {
+    IgnoredException.addIgnoredException(DistributedSystemDisconnectedException.class);
+    IgnoredException.addIgnoredException("Cause parsing to fail");
+    IgnoredException.addIgnoredException("Exception while initializing an instance");
```
Comment (Bill) [inline geode-core/src/distributedTest/java/org/apache/geode/cache30/ReconnectDUnitTest.java:1123]: Now that the (broken) regexp is no longer hiding exceptions, and you've explicitly added expectations for all of them, this test serves a good documentation for these exceptions and their role in reconnect.

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java [hunk 3]
```diff
@@ -69,6 +70,8 @@
 
   @Before
   public void setUp() throws IOException {
+    // set a property to tell membership to create a new cluster
+    System.setProperty(GMSJoinLeave.BYPASS_DISCOVERY_PROPERTY, "true");
```
Comment (Bill) [inline geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java:74]: I assume this is done to speed up the test. I like it!

But shouldn't we be restoring the system property after the test runs? I believe we have a JUnit rule for that?

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java [hunk 4]
```diff
@@ -97,16 +118,16 @@ public void constructs() {
 
     assertThatCode(() -> {
       internalLocator =
-          new InternalLocator(port, loggingSession, logFile, logWriter, securityLogWriter,
+          new InternalLocator(port, loggingSession, logFile, null, null,
```
Comment (Bill) [inline geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java:121]: My read of `InternalLocator` is that it's ok to provide `null` log writers to the constructor ✓

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java [hunk 5]
```diff
@@ -88,6 +91,24 @@ public void tearDown() {
...
+        null, bindAddress, joinCluster,
+        distributedSystemProperties, hostnameForClients, workingDirectory);
+    port = internalLocator.getPort();
+    // the locator shouldn't attempt a reconnect because it's not part of a cluster
+    internalLocator.stoppedForReconnect = true;
+    assertThat(internalLocator.attemptReconnect()).isFalse();
+    String output = FileUtils.readFileToString(logFile, Charset.defaultCharset());
+    assertThat(output).isNotEmpty();
+    assertThat(output).contains(InternalLocator.IGNORING_RECONNECT_REQUEST);
+  }
```
Comment (Bill) [inline geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java:108]: I guess we had the reconnect cases covered (in the DUnit test) and this new test is for the no-reconnect case ✓

**Review by Bill — CHANGES_REQUESTED** [review]
Only problem I see is you want to reset the system property at the end of the integration test.

File: geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java (hunk 3 above)
Comment (Bill) [reply geode-core/src/integrationTest/java/org/apache/geode/distributed/internal/InternalLocatorIntegrationTest.java:74]: I just noticed that @bschuchardt did add a `RestoreSystemProperties` rule so this is all good now

**Review by Bill — APPROVED** [review]

**Pull request merged**